### PR TITLE
Run FPC CC via External Builders for FPC chaincode

### LINKED ISSUE #249: Launch FPC Chaincode via external builder
This will allow a cleaner integration of fpc as a separate chaincode type and remove (mostly) the need of the peer cli wrapper.  It should also remove need of the proxy and sgx patch for the peer and the docker-switcheroo hack in the peer cli wrapper. 

Aspects on it:
- [x] identify config changes and semantics of scripts (and verify experimentally :-)
- [x] define fpc package format (presumably some structure as others, just with a new type fpc-c in the `type` field in `metadata.json`
- [ ] deployment model: 
  - [x] minimal version, running directly on host
  - [ ] initially just quick hack via plain local dockerand storing all state in the directories provided by peer to scripts
  - [ ] eventually consider moving to k8s so we can scale although this also raises some additional sgx questions ...). See also issue #401
  - [ ] support for chaincode-as-a-service mode (i.e., updating and testing `release.sh` script)
- [x] create the fpc-specific scripts `detect`, `build`, `release` and `run` with focus on normal mode and not yet CCaaS (i.e., the `release` script wouldn't really do anything ...).  This is mostly refactoring current code from `fabric/bin/peer.sh` to `build' and `run'.
- [x] change `core.yaml` and `fabric/bin/peer.sh` correspondingly to enable external builders and do packaging wrapper

Background:
- [main docu](https://hyperledger-fabric.readthedocs.io/en/release-2.0/cc_launcher.html)
- [related chaincode-as-a-service doc](https://hyperledger-fabric.readthedocs.io/en/release-2.0/cc_service.html)
- sample builders: integration/externalbuilders/golang in fabric repo

### PR DESCRIPTION
**Pull request by g2flyer**
**What this PR does / why we need it**:
- enable external builders in integration test config
- modify peer.sh to remove install/instantiate wrapping for fpc-CC but add a wrapper for lifecycle chaincode package
- create the external builder scripts, currently just doing essentially what the wrapper has done before, although not yet vi docker but directly on host ..

**Which issue(s) this PR fixes**:
Fixes #249, also addresses parts of #254 

**Special notes for your reviewer**:
As we do not yet have decorators and tlcc, 
- ercc registration will fail (but corresponding error is currently ignored, so corresponding TODO in peer.sh)
- stateful transactions do not work (or more precisely, transaction which want to read state; the initial write seems to work).

Also, right now there are three (potential) limitations:
- I haven't tested yet with SGX HW mode but before we have at least the decorator sorted out, it didn't make sense to already spend time on that. However, i don't think it should require changes to the external builder itself in host-run mode (docker would be a different story, see comments in `common.sh`
- the chaincode runs directly on the host. the scaffolding for running it on docker is there but getting that done does not seem to be on the critical patch and can be done later
- the peer currently does not work for non-fpc CC.  What would have to be done is described in TODOs in peer.sh but not yet implemented. Right now i don't think any of our examples or the demo was using the peer wrapper also for non-fpc cc and with the ercc refactoring, the whole ercc setup process during channel join will have to modified any way, so there didn't seem a point in implementing this now only to be thrown out soon ..

None of the integration tests are yet converted/tested and i just tested manually.  To get you a quick start with manual testing, see below what i was using as template for testing:
```
export FABRIC_LOGGING_SPEC=debug
export FABRIC_CFG_PATH=$(/bin/pwd)/integration/config/ 
export FABRIC_BIN_DIR=$(readlink -f "../../hyperledger/fabric/build/bin")
export PATH=${FABRIC_BIN_DIR}:${PATH}

./fabric/bin/ledger_shutdown.sh

CHAN_ID=mychannel
PEER=./fabric/bin/peer.sh

./fabric/bin/ledger_init.sh

# example chaincode to use ..
# - echo
LBL=echo
VER=0
CC_LANG=fpc-c
CC_PATH=examples/echo/_build/lib/
#
# - Auction
LBL=auction
VER=0
CC_LANG=fpc-c
CC_PATH=examples/auction/_build/lib/

PKG=/tmp/${LBL}.tar.gz

${PEER} lifecycle chaincode package --lang ${CC_LANG} --label ${LBL} --path ${CC_PATH} ${PKG}
${PEER} lifecycle chaincode install ${PKG}

${PEER} lifecycle chaincode queryinstalled

PKG_ID=$(${PEER} lifecycle chaincode queryinstalled | awk "/Package ID: ${LBL}/{print}" | sed -n 's/^Package ID: //; s/, Label:.*$//;p')

${PEER} lifecycle chaincode approveformyorg -C ${CHAN_ID} --package-id ${PKG_ID} --name ${LBL} --version ${VER} # --init-required # extra flag required for auction test
${PEER} lifecycle chaincode checkcommitreadiness -C ${CHAN_ID} --name ${LBL} --version ${VER}

${PEER} lifecycle chaincode commit -C ${CHAN_ID} --name ${LBL} --version ${VER}  # --init-required # extra flag required for auction test

${PEER} lifecycle chaincode querycommitted -C ${CHAN_ID}

# - echo
${PEER} chaincode invoke -C ${CHAN_ID} -n ${LBL} -c '{"Args": ["echo-1"]}'
#  should work ..

# - auction
${PEER} chaincode invoke -C ${CHAN_ID} -n ${LBL} --isInit -c '{"Args":["My Auction"]}'
# should succeed
${PEER} chaincode invoke -C ${CHAN_ID} -n ${LBL} -c '{"Function":"create", "Args": ["MyAuction1"]}' --waitForEvent
# will fail due to tlcc not being there ...

./fabric/bin/ledger_shutdown.sh

```

 
**Does this PR introduce a user-facing changes and/or breaks backward compatability?**:

The peer wrapper does not work with multiple peers anymore (strictly speaking, we never tested multi-peer in old version and the lack of support is already since #307,  but just spelling it here explicitly)

### CODE REVIEW

File: integration/config/core.yaml [hunk 1]
```diff
@@ -539,12 +539,28 @@ chaincode:
...
         #   name: descriptive-builder-name
         #   environmentWhitelist:
         #      - ENVVAR_NAME_TO_PROPAGATE_FROM_PEER
         #      - GOPROXY
+    externalBuilders:
+        # FPC Addition 0: external builder for fpc-c chaincode
+        - path: ../../fabric/externalBuilder
+          name: fpc-c
+          environmentWhitelist:
+              - FPC_HOSTING_MODE
```
Comment (bvavala) [inline integration/config/core.yaml:553]: This is nice!
Not sure about the host mode, I think it should always be docker.
But a similar variable can be used to distinguish between FPC cc and FPC architectural cc (i.e. ERCC).

File: integration/config/core.yaml.orig [hunk 2]
```diff
@@ -0,0 +1,746 @@
+# Copyright IBM Corp. All Rights Reserved.
+# Copyright Intel Corp. All Rights Reserved.
```
Comment (bvavala) [inline integration/config/core.yaml.orig:2]: (see license comment)

File: integration/config/orderer.yaml [hunk 3]
```diff
@@ -3,7 +3,7 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 #
-# Based on hyperledger/fabric/sampleconfig
```
Comment (bvavala) [inline integration/config/orderer.yaml:6]: (see license comment)

File: integration/config/msp/config.yaml [hunk 4]
```diff
@@ -3,7 +3,7 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 #
-# Based on hyperledger/fabric/sampleconfig/msp
```
Comment (bvavala) [inline integration/config/msp/config.yaml:6]: (see license comment)

File: fabric/bin/peer.sh [hunk 5]
```diff
@@ -17,298 +29,211 @@ FABRIC_SCRIPTDIR="${FPC_TOP_DIR}/fabric/bin/"
...
+    # the root of 'code.tar.gz'. (Eventually we might add reproducible build to the
+    # external builder, in which case we would stuff the related source into 'src/...'
+    # for metadata.json use the params passed to us, i.e., in particular type 'fpc-c'.
+
+    FPC_PKG_SANDBOX="/tmp/fpc-pkg-sandbox.$$"
+    try mkdir -p ${FPC_PKG_SANDBOX}
+
+    # - create code.tar.gz
+    ENCLAVE_FILE="enclave.signed.so"
+    MRENCLAVE_FILE="mrenclave"
```
Comment (bvavala) [inline fabric/bin/peer.sh:113]: Do we need this mrenclave file?
I think this is here for backward compatibility -- and it's fine.
As we move on, the mrenclave should be loaded from elsewhere -- likely ercc.

File: fabric/bin/peer.sh [hunk 6]
```diff
@@ -17,298 +29,211 @@ FABRIC_SCRIPTDIR="${FPC_TOP_DIR}/fabric/bin/"
...
+		CC_ENCLAVESOPATH="$2"
 		shift; shift
 		;;
 	    -l|--lang)
-		CC_LANG=$2
+		CC_LANG="$2"
 		shift;shift
 		;;
-	    *)
+	    -s|--sgx-mode)
```
[the remark this answers is not in the document]
Comment (g2flyer) [reply fabric/bin/peer.sh:54]: the external builder will have to know the sgx mode so the docker backend can be "instrumented" correctly with device and socket (or not), depending on mode ...

File: fabric/bin/peer.sh [hunk 7]
```diff
@@ -17,298 +29,211 @@ FABRIC_SCRIPTDIR="${FPC_TOP_DIR}/fabric/bin/"
...
     OTHER_ARGS=()
     while [[ $# > 0 ]]; do
 	case "$1" in
-	    -n|--name)
-		CC_NAME=$2;
-		shift; shift
-		;;
-	    -v|--version)
-		CC_VERSION=$2
+	    --label)
```
[the remark this answers is not in the document]
Comment (g2flyer) [reply fabric/bin/peer.sh:42]: this is a standard fabric argument of the new lifecycle commands. It's essentially just a local name for a package ...

[the remark this answers is not in the document]
Comment (bvavala) [reply fabric/bin/peer.sh:42]: never mind, I have looked at `peer chaincode`

File: fabric/bin/peer.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (bvavala) [reply fabric/bin/peer.sh:54]: let me rephrase.
Since we already have the `$SGX_MODE` env variable and CC_LANG` (possibly set to `fpc`), the flag does not seem necessary. You care care about the mode only for fpc chaincodes. So sim mode should be the default one and, if `SGX_MODE` is set, use it instead. Or am I missing something?

File: fabric/bin/peer.sh (hunk 5 above)
Comment (g2flyer) [reply fabric/bin/peer.sh:113]: The current code needs it functionally afaicr.  But in principle i also don't think we really need it. We have discussed that in the past and punted the related discussion when we refactor to the e2e secure version. So that time is certainly now (but probably not in this PR :-)

File: fabric/bin/peer.sh [hunk 8]
```diff
@@ -17,298 +29,211 @@ FABRIC_SCRIPTDIR="${FPC_TOP_DIR}/fabric/bin/"
...
+    # - 'metadata.json', a json object with 'path', 'type' and 'label' string fields
+    # - 'code.tar.gz' a gzipped tar-fil containing files
+    #    - 'src/...'
+    # as for fpc for now we will package already built artifacts 'enclave.signed.so' and
+    # 'mrenclave', we will skip 'src' and directly place the built artifacts into
+    # the root of 'code.tar.gz'. (Eventually we might add reproducible build to the
+    # external builder, in which case we would stuff the related source into 'src/...'
+    # for metadata.json use the params passed to us, i.e., in particular type 'fpc-c'.
+
+    FPC_PKG_SANDBOX="/tmp/fpc-pkg-sandbox.$$"
```
[the remark this answers is not in the document]
Comment (g2flyer) [reply fabric/bin/peer.sh:108]: ok

File: integration/config/core.yaml (hunk 1 above)
Comment (mbrandenburger) [reply integration/config/core.yaml:553]: I like the docker-less approach! I suggest we add dockerized external launcher to parking and see if there is need for it from Fabric community.

**Comment by mbrandenburger** [comment]
> * the peer currently does not work for non-fpc CC.  What would have to be done is described in TODOs in peer.sh but not yet implemented. Right now i don't think any of our examples or the demo was using the peer wrapper also for non-fpc cc and with the ercc refactoring, the whole ercc setup process during channel join will have to modified any way, so there didn't seem a point in implementing this now only to be thrown out soon ..

We have the `deployment_test.sh` that installs and runs an fpc auction and go example02 at the same time.

File: fabric/bin/peer.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (bvavala) [reply fabric/bin/peer.sh:54]: - as you mention, I was actually referring to `--lang`. So, if it's an FPC chaincode, then SIM mode should be the default, and if `SGX_MODE` is set, just use it. That's why it seems unnecessary and I have asked for a use case.
- it might make sense if in the future this will be the only place where sgx-mode is set -- so we will drop the `SGX_MODE` env variable.

This said, I'll leave it up to you.

[the remark this answers is not in the document]
Comment (g2flyer) [reply fabric/bin/peer.sh:54]: > We have the `deployment_test.sh` that installs and runs an fpc auction and go example02 at the same time.

Good point. of course we could change deployment_test.sh correspondingly but changing the peer as outlined in the TODOs is probably better (and shouldn't be fundamentally more complicated).  But maybe still do better in a separate PR (and once we have tlcc & decorators or related hacks to get the examples actually working)?

File: integration/config/core.yaml (hunk 1 above)
Comment (g2flyer) [reply integration/config/core.yaml:553]: well, from a both deployment and security perspective docker is better.  But it's certainly not on the critical path ...

File: fabric/bin/peer.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (g2flyer) [reply fabric/bin/peer.sh:54]: ok, if you put it like that, i leave it as-is :-)

**Comment by mbrandenburger** [comment]
@g2flyer please rebase/autosquash! thanks!

**Comment by mbrandenburger** [comment]
While working on ercc decorator I added the `try` to `__setup` invocation as suggested in the TODO.

Works fine with echo but auction complains that it cannot invoke `__setup` before init is called.

```
2020-04-07 10:22:15.935 UTC [chaincodeCmd] chaincodeInvokeOrQuery -> DEBU 006 ESCC invoke result: response:<status:500 message:"error in simulation: failed to execute transaction dd79ff5abad8ab412fd91757175f316a10b636fea438d0e7229ca5f156810b88: invalid invocation: chainc$
de 'auction' has not been initialized for this version, must call as init first" >
Error: endorsement failure during invoke. response: status:500 message:"error in simulation: failed to execute transaction dd79ff5abad8ab412fd91757175f316a10b636fea438d0e7229ca5f156810b88: invalid invocation: chaincode 'auction' has not been initialized for this version,
must call as init first"
peer.sh: test failed: /project/src/github.com/hyperledger/fabric/build/bin/peer chaincode invoke -o localhost:7050 -C mychannel -n auction -c {"Args":["__setup", "ercc"]} --waitForEvent
```

**Comment by mbrandenburger** [comment]
> * stateful transactions do not work (or more precisely, transaction which want to read state; the initial write seems to work).

tested with hello_world example and get/put_state works. However, notice that ecc currently uses tlcc mock and therefore we don't get any tlcc verification errors.
